cleverhans/picklable_model.py

```diff
@@ -0,0 +1,370 @@
+        self.__dict__.update(locals())
+        del self.self
+        super(Conv2D, self).__init__(**kwargs)
+
+    def set_input_shape(self, input_shape):
+        batch_size, rows, cols, input_channels = input_shape
+        kernel_shape = tuple(self.kernel_shape) + (input_channels,
+                                                   self.output_channels)
+        assert len(kernel_shape) == 4
+        assert all(isinstance(e, int) for e in kernel_shape), kernel_shape
+        init = tf.random_normal(kernel_shape, dtype=tf.float32)
+        init = self.init_scale * init / tf.sqrt(1e-7 +
+                                                tf.reduce_sum(tf.square(init),
+                                                              axis=(0, 1, 2)))
+        self.kernels = PV(init)
+        if self.use_bias:
+            self.b = PV(np.zeros((self.output_channels,)).astype('float32'))
+        orig_input_batch_size = input_shape[0]
+        input_shape = list(input_shape)
+        input_shape[0] = 1
+        dummy_batch = tf.zeros(input_shape)
+        dummy_output = self.fprop(dummy_batch)
+        output_shape = [int(e) for e in dummy_output.get_shape()]
+        output_shape[0] = 1
+        self.output_shape = tuple(output_shape)
+
+    def fprop(self, x):
+        out = tf.nn.conv2d(x, self.kernels.var,
+                           (1,) + tuple(self.strides) + (1,), self.padding)
+        if self.use_bias:
+            out = out + self.b.var
+        return out
+
+    def get_params(self):
+        out = [self.kernels.var]
+        if self.use_bias:
+            out.append(self.b.var)
+        return out
+
+
+class ReLU(Layer):
+
+    def __init__(self, leak=0., **kwargs):
+        super(ReLU, self).__init__(**kwargs)
+        self.leak = leak
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        out = tf.nn.relu(x)
+        if self.leak != 0.0:
+            out = out - self.leak * tf.nn.relu(-x)
+        return out
+
+    def get_params(self):
+        return []
+
+
+class Sigmoid(Layer):
+
+    def __init__(self, **kwargs):
+        super(Sigmoid, self).__init__(**kwargs)
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        return tf.nn.sigmoid(x)
+
+    def get_params(self):
+        return []
+
+
+class Tanh(Layer):
+
+    def __init__(self, **kwargs):
+        super(Tanh, self).__init__(**kwargs)
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        return tf.nn.tanh(x)
+
+    def get_params(self):
+        return []
+
+
+class LeakyReLU(Layer):
```

Comment:
Why are there two implementations of leaky ReLU? (this one and through `leak=0.` in ReLU class)?

Reply:
I changed LeakyReLU to have nothing but a constructor supplying a default leak amount

Reply:
Thanks!